policy/reviewer/reviewer.go

```diff
@@ -106,54 +125,70 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
 
 	for _, child := range pendingLeafNodes {
 		allUsers := make(map[string]struct{})
-		shoveIntoMap(allUsers, child.ReviewRequestRule.Users)
+
+		if len(child.ReviewRequestRule.Users) > 0 {
+			for _, user := range child.ReviewRequestRule.Users {
+				allUsers[user] = struct{}{}
+			}
+		}
 
 		if len(child.ReviewRequestRule.Teams) > 0 {
 			teamMembers, err := selectTeamMembers(prctx, child.ReviewRequestRule.Teams, r)
 			if err != nil {
 				logger.Warn().Err(err).Msgf("Unable to get member listing for teams, skipping team member selection")
 			}
-			shoveIntoMap(allUsers, teamMembers)
+			for _, user := range teamMembers {
+				allUsers[user] = struct{}{}
+			}
 		}
 
 		if len(child.ReviewRequestRule.Organizations) > 0 {
 			orgMembers, err := selectOrgMembers(prctx, child.ReviewRequestRule.Organizations, r)
 			if err != nil {
 				logger.Warn().Err(err).Msg("Unable to get member listing for org, skipping org member selection")
 			}
-			shoveIntoMap(allUsers, orgMembers)
+			for _, user := range orgMembers {
+				allUsers[user] = struct{}{}
+			}
 		}
 
-		allCollaboratorPermissions, err := prctx.RepositoryCollaborators()
+		collaboratorsToConsider := make(map[string]string)
+		allCollaborators, err := prctx.RepositoryCollaborators()
 		if err != nil {
 			return nil, errors.Wrap(err, "Unable to list repository collaborators")
 		}
 
-		if child.ReviewRequestRule.Admins {
-			var repoAdmins []string
-			for user, perm := range allCollaboratorPermissions {
-				if perm == common.GithubAdminPermission {
-					repoAdmins = append(repoAdmins, user)
+		if child.ReviewRequestRule.WriteCollaborators {
+			for user, perm := range allCollaborators {
+				if perm == common.GithubWritePermission {
+					allUsers[user] = struct{}{}
 				}
 			}
-			shoveIntoMap(allUsers, repoAdmins)
 		}
 
-		if child.ReviewRequestRule.WriteCollaborators {
-			var repoCollaborators []string
-			for user, perm := range allCollaboratorPermissions {
-				if perm == common.GithubWritePermission {
-					repoCollaborators = append(repoCollaborators, user)
-				}
+		// When admins are selected for review, only collect the desired set of admins instead of
+		// everyone, which includes org admins
+		if !child.ReviewRequestRule.Admins {
+			// When not looking for admins, we want to check with all possible collaborators
+			collaboratorsToConsider = allCollaborators
```

Comment:
These names are better now, but seeing it in the new limited context makes me think the distinction is not necessary at all, and is actually incorrect.

The `allUsers` set is what we actually use to select reviewers, while `collboratorsToConsider` is only used to figure out which elements of `allUsers` are collaborators. In the admin case, we add the team admins to `allUsers` but then we additionally limit the set of collaborators to admins as well. When an approval rule allows both admins and other conditions, I think this artificially limits the set of reviewers to only the admins. And when the rule only allows admins, the admin/not-admin filtering is performed by `selectAdmins`, so the additional filtering at this level is redundant.

Am I missing something?

Reply:
I think you're right. I added the distinction previously when trying to find the set of admins and then remove certain users like org owners. Now that we're going with a purely additive approach, we can simplify and remove this bug.